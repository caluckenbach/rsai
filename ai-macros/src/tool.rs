use proc_macro2::TokenStream;
use quote::quote;
use syn::{ItemFn, Result, FnArg, Pat, Type, Attribute};

pub fn tool_impl(attr: TokenStream, item: TokenStream) -> Result<TokenStream> {
    let _ = attr; // Currently unused, could be used for tool configuration
    
    let input = syn::parse2::<ItemFn>(item)?;
    let fn_name = &input.sig.ident;
    let fn_name_str = fn_name.to_string();
    
    // Extract function description and parameter descriptions from doc comments
    let (description, param_descriptions) = extract_doc_comment_and_params(&input.attrs);
    
    // Parse function parameters
    let params = parse_parameters(&input.sig.inputs, &param_descriptions)?;
    
    // Generate JSON schema for parameters
    let schema = generate_parameter_schema(&params)?;
    
    // Generate a function that returns the Tool struct
    let tool_fn_name = quote::format_ident!("{}_tool", fn_name);
    
    // Generate the complete implementation
    let expanded = quote! {
        #input
        
        pub fn #tool_fn_name() -> ai_rs::core::types::Tool {
            ai_rs::core::types::Tool {
                name: #fn_name_str.to_string(),
                description: #description,
                parameters: #schema,
                strict: Some(true),
            }
        }
    };
    
    Ok(expanded)
}

fn extract_doc_comment_and_params(attrs: &[Attribute]) -> (TokenStream, std::collections::HashMap<String, String>) {
    let doc_strings: Vec<String> = attrs
        .iter()
        .filter_map(|attr| {
            if attr.path().is_ident("doc") {
                if let syn::Meta::NameValue(meta) = &attr.meta {
                    if let syn::Expr::Lit(expr_lit) = &meta.value {
                        if let syn::Lit::Str(lit_str) = &expr_lit.lit {
                            return Some(lit_str.value().trim().to_string());
                        }
                    }
                }
            }
            None
        })
        .collect();
    
    let mut description_lines = Vec::new();
    let mut param_descriptions = std::collections::HashMap::new();
    
    for line in doc_strings {
        // Check if this line describes a parameter (format: "param_name: description")
        if let Some(colon_pos) = line.find(':') {
            let param_name = line[..colon_pos].trim();
            let param_desc = line[colon_pos + 1..].trim();
            
            // Only treat it as a parameter description if the param name looks like an identifier
            if param_name.chars().all(|c| c.is_alphanumeric() || c == '_') && !param_name.is_empty() {
                param_descriptions.insert(param_name.to_string(), param_desc.to_string());
                continue;
            }
        }
        
        // Otherwise, it's part of the function description
        description_lines.push(line);
    }
    
    let description = if description_lines.is_empty() {
        quote! { None }
    } else {
        let description = description_lines.join(" ");
        quote! { Some(#description.to_string()) }
    };
    
    (description, param_descriptions)
}

struct Parameter {
    name: String,
    ty: Type,
    description: Option<String>,
    required: bool,
}

fn parse_parameters(
    inputs: &syn::punctuated::Punctuated<FnArg, syn::token::Comma>,
    param_descriptions: &std::collections::HashMap<String, String>,
) -> Result<Vec<Parameter>> {
    let mut params = Vec::new();
    
    for arg in inputs {
        match arg {
            FnArg::Receiver(_) => {
                return Err(syn::Error::new_spanned(arg, "tool functions cannot have self parameter"));
            }
            FnArg::Typed(pat_type) => {
                let name = match &*pat_type.pat {
                    Pat::Ident(pat_ident) => pat_ident.ident.to_string(),
                    _ => return Err(syn::Error::new_spanned(&pat_type.pat, "only simple identifiers are supported for parameters")),
                };
                
                // Get parameter description from docstring parsing
                let description = param_descriptions.get(&name).cloned();
                
                // Check if type is Option<T>
                let (ty, required) = match &*pat_type.ty {
                    Type::Path(type_path) => {
                        let segments = &type_path.path.segments;
                        if segments.len() == 1 && segments[0].ident == "Option" {
                            // Extract inner type from Option<T>
                            if let syn::PathArguments::AngleBracketed(args) = &segments[0].arguments {
                                if let Some(syn::GenericArgument::Type(inner_ty)) = args.args.first() {
                                    (inner_ty.clone(), false)
                                } else {
                                    ((*pat_type.ty).clone(), true)
                                }
                            } else {
                                ((*pat_type.ty).clone(), true)
                            }
                        } else {
                            ((*pat_type.ty).clone(), true)
                        }
                    }
                    _ => ((*pat_type.ty).clone(), true),
                };
                
                params.push(Parameter {
                    name,
                    ty,
                    description,
                    required,
                });
            }
        }
    }
    
    Ok(params)
}


fn generate_parameter_schema(params: &[Parameter]) -> Result<TokenStream> {
    if params.is_empty() {
        return Ok(quote! {
            ::serde_json::json!({
                "type": "object",
                "properties": {},
                "required": [],
                "additionalProperties": false
            })
        });
    }

    let properties: Vec<_> = params.iter().map(|param| {
        let name = &param.name;
        let type_str = type_to_json_type(&param.ty).unwrap_or("string");
        
        if let Some(desc) = &param.description {
            quote! {
                (#name, ::serde_json::json!({
                    "type": #type_str,
                    "description": #desc
                }))
            }
        } else {
            quote! {
                (#name, ::serde_json::json!({
                    "type": #type_str
                }))
            }
        }
    }).collect();

    let required_params: Vec<_> = params.iter()
        .filter(|p| p.required)
        .map(|p| &p.name)
        .collect();
    
    Ok(quote! {
        {
            let mut properties = ::serde_json::Map::new();
            #(
                properties.insert(#properties.0.to_string(), #properties.1);
            )*
            
            ::serde_json::json!({
                "type": "object",
                "properties": properties,
                "required": [#(#required_params),*],
                "additionalProperties": false
            })
        }
    })
}

fn type_to_json_type(ty: &Type) -> Result<&'static str> {
    match ty {
        Type::Path(type_path) => {
            let ident = &type_path.path.segments.last()
                .ok_or_else(|| syn::Error::new_spanned(ty, "empty type path"))?
                .ident;
            
            match ident.to_string().as_str() {
                "String" | "str" => Ok("string"),
                "i8" | "i16" | "i32" | "i64" | "i128" | "isize" |
                "u8" | "u16" | "u32" | "u64" | "u128" | "usize" => Ok("integer"),
                "f32" | "f64" => Ok("number"),
                "bool" => Ok("boolean"),
                "Vec" => Ok("array"),
                _ => Ok("object"), // Default to object for complex types
            }
        }
        _ => Ok("object"),
    }
}