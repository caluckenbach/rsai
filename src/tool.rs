use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::HashMap;

// Core tool definition
#[derive(Debug, Clone)]
pub struct Tool {
    pub name: String,
    pub description: String,
    pub parameters: Parameters,
    pub execute: fn(HashMap<String, String>) -> Value,
}

// Parameter definitions for tools
#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct Parameters {
    #[serde(rename = "type")]
    pub param_type: String,
    pub properties: HashMap<String, Property>,
    pub required: Vec<String>,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct Property {
    #[serde(rename = "type")]
    pub property_type: String,
    pub description: String,
}

// Function declaration for LLM API
#[derive(Serialize, Deserialize, Debug)]
pub struct FunctionDeclaration {
    pub name: String,
    pub description: String,
    pub parameters: Parameters,
}

// Helper function to create a new tool
pub fn tool(
    name: &str,
    description: &str,
    parameters: HashMap<String, (String, String)>,
    required: Vec<&str>,
    execute: fn(HashMap<String, String>) -> Value,
) -> Tool {
    let mut properties = HashMap::new();
    for (key, (property_type, description)) in parameters {
        properties.insert(
            key.to_string(),
            Property {
                property_type,
                description,
            },
        );
    }

    Tool {
        name: name.to_string(),
        description: description.to_string(),
        parameters: Parameters {
            param_type: "object".to_string(),
            properties,
            required: required.iter().map(|&s| s.to_string()).collect(),
        },
        execute,
    }
}

// Process a function call by finding the appropriate tool and executing it
pub fn process_function_call(
    tools: &[Tool],
    function_name: &str,
    args: &Value,
) -> Option<Value> {
    let tool = tools.iter().find(|t| t.name == function_name)?;

    // Convert args to HashMap<String, String>
    let mut params = HashMap::new();
    if let Some(obj) = args.as_object() {
        for (key, value) in obj {
            if let Some(value_str) = value.as_str() {
                params.insert(key.clone(), value_str.to_string());
            }
        }
    }

    Some((tool.execute)(params))
}