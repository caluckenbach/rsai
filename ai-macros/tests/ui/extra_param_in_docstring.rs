// Mock the ai_rs types for compile testing
mod ai_rs {
    pub mod core {
        pub mod types {
            use std::future::Future;
            use std::pin::Pin;
            
            #[derive(Debug, Clone, PartialEq)]
            pub struct Tool {
                pub name: String,
                pub description: Option<String>,
                pub parameters: serde_json::Value,
                pub strict: Option<bool>,
            }
            
            pub type BoxFuture<'a, T> = Pin<Box<dyn Future<Output = T> + Send + 'a>>;
        }
        
        pub mod error {
            #[derive(Debug)]
            pub enum LlmError {
                ToolExecution {
                    message: String,
                    source: Option<Box<dyn std::error::Error + Send + Sync>>,
                },
            }
        }
        
        pub trait ToolFunction: Send + Sync {
            fn schema(&self) -> types::Tool;
            fn execute<'a>(&'a self, params: serde_json::Value) -> types::BoxFuture<'a, Result<serde_json::Value, error::LlmError>>;
        }
    }
}

use ai_macros::tool;

#[tool]
/// Function with extra parameter in docstring
/// param1: Valid parameter description
/// nonexistent: This parameter doesn't exist in the function
fn function_with_extra_param(param1: String) -> String {
    param1
}

fn main() {}