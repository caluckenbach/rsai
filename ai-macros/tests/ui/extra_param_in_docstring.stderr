error: Parameter 'nonexistent' found in docstring but not in function parameters
  --> tests/ui/extra_param_in_docstring.rs:42:1
   |
42 | fn function_with_extra_param(param1: String) -> String {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
